### ISSUE TITLE
Reconsider making Numpy the default. 

### ISSUE THREAD

**Issue by MuawizChaudhary** [kymatio/kymatio#806]
I see very few reasons for Numpy to be the default back/frontend of the Kymatio software package. 

On the [kymat.io](https://www.kymat.io/) website we state 
 
> Use Kymatio if you need a library that:
> - integrates wavelet scattering in a deep learning architecture, and
> - runs seamlessly on CPU and GPU hardware, with major deep learning APIs, such as PyTorch and TensorFlow.

A large amount of our presentation and motivation behind Kymatio has included the following points; seamless integration within a deep learning pipeline, differentiable wavelets, capable of running on GPU, and implementation of scattering in a modern deep learning framework. 

To change the default backend to Numpy is henious in this context; Numpy is rarely used as a part of a deep learning architecture, owning to its lack of differentiablity and lack of utilization of the GPU. Furthermore, I believe a large majority of the research utilizing Kymatio largely uses the Torch back/frontend of our software. 

I highly reccomend reconsidering making Numpy the default back/frontend. This role should instead be relegated to a proper deep learning library, such as Pytorch (which is the current default).

**Comment by MuawizChaudhary** [comment]
Backwards compatability is indeed a large problem with Pytorch. 

Aside from Torch, Torchvision, TorchAudio, can you note for me which dependencies would be removed if Numpy became the default? 

Google Colab is a very popular environment for budding deep learners, novice data scienctists, and students (my undergrad encoraged us to use Google Colab). Torch is already preinstalled on the python environments there. 

Of course, users on their local systems will have to install Torch. 

We should also keep in mind the portion of our users who are researchers and utilizing this package. As developers we should be focusing on giving our users high impact features they are very likely to use and take advantage of. 

For example, in this JTFS PR we are focusing on getting the Numpy version out first. Ignoring the [recent work+repo relese by your group with JTFS and unsupervised learning,](https://github.com/rastegah/kymatio-jtfs) this development process focusing on Numpy first could delay getting our users a stable version of a GPU enabled, differentiable scattering. 

I will be radical and state that infact we should not support Numpy in the main library (that is, we fork it!) and that we should be totally focused instead on supporting Tensorflow, Jax, and Torch.   ;)

**Comment by lostanlen** [comment]
@MuawizChaudhary it's a question of long-term evolution.

> Aside from Torch, Torchvision, TorchAudio, can you note for me which dependencies would be removed if Numpy became the default?

Those. But Torch itself (1) has a mind-boggling number of non-Python dependencies, (2) comes in many flavors depending on platforms, and (3) has undergone tremendous breaking changes over the past few years.

This means that if a newcomer who's unfamiliar with PyTorch installs Kymatio, runs our tutorials, and encounters a runtime error, we will have to say: what's your conda/pip environment? what's your platform? what's your pytorch version?
Those three daunting questions alone might shoo away the person who reported the problem and we might never be able to help them. Worse, it will mean we'll spend more time on dependency resolution and less on actual Kymatio code. Of course, a more experienced user (e.g., researcher) will readily answer those questions; but that demographic is also the one that's most capable at helping themself.

Keep in mind that the SciPy/NumPy stack has ~20 years of experience of working with complex numbers and linear algebra. PyTorch. According to PyTorch documentation: (https://pytorch.org/docs/stable/complex_numbers.html)

> Complex tensors is a beta feature and subject to change.

So there is no possible comparison in terms of maturity and stability.


> Google Colab is a very popular environment for budding deep learners, novice data scientists, and students (my undergrad encouraged us to use Google Colab). Torch is already preinstalled on the python environments there.

Tying ourselves to a corporate environment is a seductive but dangerous idea.
But again, don't we have the best of both worlds with
```
from kymatio.torch import Scattering1D, Scattering2D
```
? 
Which should work out of the box in Google Colab?

> For example, in this JTFS PR, we are focusing on getting the Numpy version out first. Ignoring the [recent work+repo relese by your group with JTFS and unsupervised learning,](https://github.com/rastegah/kymatio-jtfs) this development process focusing on Numpy first delays getting our users a stable version of a GPU enabled, differentiable scattering.

Take that to @OverLordGoldDragon who wrote the code.

> I will be radical and state that infact we should not support Numpy in the main library (that is, we fork it!) and that we should be totally focused instead on supporting Tensorflow, Jax, and Torch. ;)

Frankly, i could let myself convinced. But i am concerned about obsolescence.
Here's an anecdote: between 2013 and 2015, Joakim and i wrote implementations of JTFS in MATLAB. These implementations still run fine to this day (2022).
In January 2016, i wrote a neural network implementation in Keras/Theano. The implementation was already no longer compatible with latest stable Keras when i presented my paper in the Summer of 2016.

So yes let's think about the researchers of the current day, but also those of the day after that.

Just my party-spoiler thoughts. If you think the pros outweight the cons, and you bet that PyTorch will become as stable as is current NumPy, _and_ (most importantly) if you're willing to take the responsibility of helping newcomers troubleshoot their PyTorch-based Kymatio installations, then by all means, let me retract what i wrote above and let's boldly go (back) to PyTorch as default backend.

This is something worth discussing viva voce on May 19 :)

**Comment by janden** [comment]
How about not having a default at all? Just have the user import from `kymatio.*` depending on their needs? If we do want a default, however, I agree with @lostanlen that it should be NumPy, for the reasons he stated.
